File: frontend/src/lib/crypto.ts
import crypto from 'crypto';

// Derives a 32-byte key from the ENCRYPTION_KEY environment variable.
const getEncryptionKey = (): Buffer => {
  const secret = process.env.ENCRYPTION_KEY;
  if (!secret) {
    if (process.env.NODE_ENV === 'production') {
      throw new Error('ENCRYPTION_KEY environment variable is required in production.');
    }
    // Fallback key for development mode
    return crypto.scryptSync('development_fallback_key_receptify', 'salt', 32);
  }
  // Ensure the key is exactly 32 bytes (via scrypt derivation or simple padding/trimming)
  return Buffer.from(secret.padEnd(32, '0').slice(0, 32));
};

const IV_LENGTH = 12; // For GCM

export function encrypt(text: string): string {
  const key = getEncryptionKey();
  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);

  let encrypted = cipher.update(text, 'utf8', 'hex');
  encrypted += cipher.final('hex');

  const authTag = cipher.getAuthTag();

  // Return a colon-separated string: iv:authTag:ciphertext
  return `${iv.toString('hex')}:${authTag.toString('hex')}:${encrypted}`;
}

export function decrypt(encryptedText: string): string {
  const parts = encryptedText.split(':');
  if (parts.length !== 3) {
    throw new Error('Invalid encrypted text format.');
  }

  const iv = Buffer.from(parts[0], 'hex');
  const authTag = Buffer.from(parts[1], 'hex');
  const encrypted = parts[2]; // Keep as string (hex format)

  const key = getEncryptionKey();
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);

  decipher.setAuthTag(authTag);

  let decrypted = decipher.update(encrypted, 'hex', 'utf8');
  decrypted += decipher.final('utf8');

  return decrypted;
}

File: backend/receptify/crypto.py
import binascii
from Cryptodome.Cipher import AES
from decouple import config

def get_encryption_key() -> bytes:
    secret = config('ENCRYPTION_KEY', default='')
    if not secret:
        # Fallback key matching TypeScript: crypto.scryptSync('development_fallback_key_receptify', 'salt', 32)
        # In Python, we can do scrypt derivation:
        import hashlib
        try:
            return hashlib.scrypt(
                password=b"development_fallback_key_receptify",
                salt=b"salt",
                n=16384,
                r=8,
                p=1,
                dklen=32
            )
        except AttributeError:
            from Cryptodome.Protocol.KDF import scrypt
            return scrypt(
                password=b"development_fallback_key_receptify",
                salt=b"salt",
                key_len=32,
                N=16384,
                r=8,
                p=1
            )
    
    # Ensure key is exactly 32 bytes via padding or slicing (matches TypeScript logic)
    padded_secret = secret.ljust(32, '0')[:32]
    return padded_secret.encode('utf-8')

def encrypt(text: str) -> str:
    key = get_encryption_key()
    cipher = AES.new(key, AES.MODE_GCM)
    ciphertext, tag = cipher.encrypt_and_digest(text.encode('utf-8'))
    
    iv_hex = binascii.hexlify(cipher.nonce).decode('utf-8')
    tag_hex = binascii.hexlify(tag).decode('utf-8')
    ciphertext_hex = binascii.hexlify(ciphertext).decode('utf-8')
    
    return f"{iv_hex}:{tag_hex}:{ciphertext_hex}"

def decrypt(encrypted_text: str) -> str:
    parts = encrypted_text.split(':')
    if len(parts) != 3:
        raise ValueError("Invalid encrypted text format.")
    
    iv = binascii.unhexlify(parts[0])
    tag = binascii.unhexlify(parts[1])
    ciphertext = binascii.unhexlify(parts[2])
    
    key = get_encryption_key()
    cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
    decrypted_bytes = cipher.decrypt_and_verify(ciphertext, tag)
    
    return decrypted_bytes.decode('utf-8')
